### PR TITLE
Adding a skeleton for storing Spark Magic events to MDS

### PR DESCRIPTION
**Pull request by MohamedElKamhawy**

### CODE REVIEW

**Comment by MohamedElKamhawy** [comment]
It's my bad, I just committed without adding the pull request.

**Comment by msftristew** [comment]
Talked f2f

File: remotespark/utils/configuration.py [hunk 1]
```diff
@@ -138,6 +138,13 @@ def logging_config():
         }
     }
 
+@_override
+def events_config():
+    return {
+        "local": 1,
```
Comment (msftristew) [inline remotespark/utils/configuration.py]: Unnecessary

File: remotespark/utils/jupytereventsfilehandler.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from remotespark.utils.filesystemreaderwriter import FileSystemReaderWriter
+
+
+class EventsFileHandler(logging.FileHandler):
+    """The default logging handler used by the magics; this behavior can be overridden by modifying the config file"""
+    def __init__(self):
+        # Simply invokes the behavior of the superclass, but sets the filename keyword argument if it's not already set.
+
+        magics_home_path = get_magics_home_path()
+        events_folder_name = "events"
```
Comment (msftristew) [inline remotespark/utils/jupytereventsfilehandler.py]: No need to create a new file, just use the logger

File: remotespark/utils/jupytereventsfilehandler.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        events_file_name = "events{}.log".format(get_instance_id())
+        directory = FileSystemReaderWriter(join_paths(magics_home_path, events_folder_name))
+        directory.ensure_path_exists()
+        events_file = join_paths(directory.path, events_file_name)
+
+    def emit(self, **kwargs):
```
Comment (msftristew) [inline remotespark/utils/jupytereventsfilehandler.py]: API needs to be more granular

File: remotespark/utils/jupyterevents.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            # Create the local file for the events
+        else:
+            # Load the hdinsighteventshandler class
+            self.local = 0
+            self._import(conf.events_config()["handler"])
+
+    def _import(self, name):
+        components = name.split('.')
+        mod = __import__(components[0])
+        for comp in components[1:]:
```
Comment (msftristew) [inline remotespark/utils/jupyterevents.py]: ?

File: remotespark/utils/jupyterevents.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# Read the config file that will return the filehandler
+# file handler will be one of two local file handler or hdinsighteventshandler
+
+import remotespark.utils.configuration as conf
+
+
+class JupyterEvents:
```
Comment (msftristew) [inline remotespark/utils/jupyterevents.py]: This name isn't really accurate

File: remotespark/utils/jupytereventsfilehandler.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import logging
+
+from remotespark.utils.utils import join_paths, get_magics_home_path, get_instance_id
+from remotespark.utils.filesystemreaderwriter import FileSystemReaderWriter
+
+
+class EventsFileHandler(logging.FileHandler):
```
Comment (msftristew) [inline remotespark/utils/jupytereventsfilehandler.py]: Shouldn't use logging infrastructure

**Comment by msftristew** [comment]
In general I would recommend against submitting incomplete PR's that are not ready to be merged.

File: remotespark/utils/configuration.py (hunk 1 above)
Comment (aggFTW) [inline remotespark/utils/configuration.py]: What does `local` mean @MohamedElKamhawy?

File: remotespark/utils/jupyterevents.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# Read the config file that will return the filehandler
+# file handler will be one of two local file handler or hdinsighteventshandler
```
Comment (aggFTW) [inline remotespark/utils/jupyterevents.py]: No mention of hdinsight in this repo please. Just say that the file handler will be one of N implementations, default one being X.

File: remotespark/utils/jupyterevents.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    def _import(self, name):
+        components = name.split('.')
+        mod = __import__(components[0])
+        for comp in components[1:]:
+            mod = getattr(mod, comp)
+        return mod
+
+    def fire_event(self, **kwargs):
+        # Store the events
+        # Based on the config file which wil control which
```
Comment (aggFTW) [inline remotespark/utils/jupyterevents.py]: Where's the rest of the file?

File: remotespark/utils/jupyterevents.py (hunk 4 above)
Comment (aggFTW) [inline remotespark/utils/jupyterevents.py]: I also don't get what's going on here.

File: remotespark/utils/jupyterevents.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import remotespark.utils.configuration as conf
+
+
+class JupyterEvents:
+    def __init__(self):
+
+        if conf.events_config()["local"] == 1:
+            # Create a local events storage
+            self.local = 1
```
Comment (aggFTW) [inline remotespark/utils/jupyterevents.py]: How will this be used?

File: remotespark/utils/jupyterevents.py (hunk 5 above)
Comment (aggFTW) [inline remotespark/utils/jupyterevents.py]: What is this class's responsibility?

File: remotespark/utils/jupytereventsfilehandler.py (hunk 6 above)
Comment (aggFTW) [inline remotespark/utils/jupytereventsfilehandler.py]: +1

**Comment by MohamedElKamhawy** [comment]
@aggFTW , @msftristew : I added the main skeleton (not the whole code yet), please have a look at it. Also, I'm still playing with livysession, this is not the final version.

Give an attention for the sparkevents & eventshandler files, please.

File: remotespark/livyclientlib/livysessionstate.py [hunk 10]
```diff
@@ -39,5 +44,5 @@ def connection_string(self):
         return self._connection_string
 
     def to_dict(self):
-        return {"id": self.session_id, "kind": self.kind, "sqlcontext": self.sql_context_created,
+        return {"GUID": self.session_guid, "id": self.session_id, "kind": self.kind, "sqlcontext": self.sql_context_created,
```
Comment (msftristew) [inline remotespark/livyclientlib/livysessionstate.py]: why is GUID in all caps when the others aren't?

File: remotespark/utils/configuration.py [hunk 11]
```diff
@@ -144,6 +144,12 @@ def logging_config():
         }
     }
 
+@_override
+def events_config():
+    return {
+        "handler": "remotespark.utils.jupytereventfilehandler"
```
Comment (msftristew) [inline remotespark/utils/configuration.py]: * Don't put the string `Jupyter` in any file names (remember, these events have nothing to do with Jupyter)

* Remember that the default handler is not a "file" handler

File: remotespark/utils/sparkevents.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import remotespark.utils.configuration as conf
+
+
+class JupyterEvents:
```
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: Change the name from JupyterEvents

Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: There is a missing change!
I'll pull a new request.

File: remotespark/utils/eventshandler.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    def __init__(self):
+        self.logger = Log("EventsHandler")
+
+    def store_event(self, **kwargs):
+        """
+        Storing the Event details using the logger.
+        """
+        event_line = ""
+        for key, value in kwargs.iteritems():
+            event_line += "{} = {}\n".format(key, value)
```
Comment (msftristew) [inline remotespark/utils/eventshandler.py]: The key-value pairs should all be on the same line.

**Comment by MohamedElKamhawy** [comment]
I have a problem with Git so reverted everything and wrote it back.
Will make sure that everything is right then will have a new PR.

**Comment by msftristew** [comment]
Mohamed, also be sure to run unit tests before pushing.

File: remotespark/livyclientlib/livysession.py [hunk 14]
```diff
@@ -50,8 +51,10 @@ def __init__(self, http_client, properties, ipython_display,
         self._status_sleep_seconds = status_sleep_seconds
         self._statement_sleep_seconds = statement_sleep_seconds
         self._create_sql_context_timeout_seconds = create_sql_context_timeout_seconds
+        session_guid = uuid.uuid4()
 
-        self._state = LivySessionState(session_id, self._http_client.connection_string,
+
+        self._state = LivySessionState(session_guid, session_id, self._http_client.connection_string,
```
Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: Is it important that the GUID be in the session state?

File: remotespark/utils/eventshandler.py [hunk 15]
```diff
@@ -0,0 +1,15 @@
+from remotespark.utils.log import Log
+
+
+class EventsHandler:
+    def __init__(self):
+        self.logger = Log("EventsHandler")
+
+    def store_event(self, **kwargs):
```
Comment (msftristew) [inline remotespark/utils/eventshandler.py]: Why `**kwargs` rather than `**args`?

Mohamed, remember that you have to find a way to order the dimensions (so keyword arguments don't work)

File: remotespark/utils/eventshandler.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+        """
+        Storing the Event details using the logger.
+        """
+        event_line = ""
+        for key, value in kwargs.iteritems():
+            event_line += "{}: {} ".format(key, value)
```
Comment (msftristew) [inline remotespark/utils/eventshandler.py]: Try: `";".join("{}: {}".format(key, value) for key, value in kwargs.iteritems()`

File: remotespark/utils/sparkevents.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+        self.handler = events_handler()
+
+    def _import_handler(self, name):
+        """
+        Importing the handler based on the passed parameter.
+        """
+        components = name.split('.')
+        mod = __import__(components[0])
+        for comp in components[1:]:
+            mod = getattr(mod, comp)
```
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: What does this do?

File: remotespark/utils/sparkevents.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+
+class SparkEvents:
+    def __init__(self):
+        """
+        Create an instance from the handler mentioned in the config file.
+        """
+        events_handler = self._import_handler(conf.events_config()["handler"])
+        self.handler = events_handler()
+
+    def _import_handler(self, name):
```
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: How is this API supposed to be used?

File: remotespark/utils/sparkevents.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+        """
+        Importing the handler based on the passed parameter.
+        """
+        components = name.split('.')
+        mod = __import__(components[0])
+        for comp in components[1:]:
+            mod = getattr(mod, comp)
+        return mod
```
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: This doesn't match up with the design we discussed in our meeting.  What is returned from this method?  I thought this class was supposed to create a handler and use it in the `emit_*` methods.

Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: We decided to leave this decision for me, remember? And since it's going be a very tiny number of code lines. I found it's better to encapsulate it in a method.

Comment (msftristew) [inline remotespark/utils/sparkevents.py]: NVM, it looks like I misunderstood. This is OK.

File: remotespark/utils/sparkevents.py (hunk 18 above)
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: No reason for this to be its own method -- just put this in `__init__`

File: remotespark/utils/sparkevents.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
+import remotespark.utils.configuration as conf
+
+
+class SparkEvents:
+    def __init__(self):
+        """
+        Create an instance from the handler mentioned in the config file.
+        """
+        events_handler = self._import_handler(conf.events_config()["handler"])
```
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: There is no reason for `events_config` to be a dictionary with a `"handler"` element -- just make it a string

Comment (msftristew) [inline remotespark/utils/sparkevents.py]: Maybe call it `events_handler_class` instead of `events_config`

File: remotespark/utils/eventshandler.py (hunk 15 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/eventshandler.py]: I see that order doesn't matter in case of [file] EventsHandler.

File: remotespark/utils/sparkevents.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        self.handler = events_handler()
+
+    def _import_handler(self, name):
+        """
+        Importing the handler based on the passed parameter.
+        """
+        components = name.split('.')
+        mod = __import__(components[0])
```
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: I'm reading online that `__import__` is an internal implementation detail.  Can you use `importlib` instead?

File: remotespark/utils/eventshandler.py (hunk 15 above)
Comment (msftristew) [inline remotespark/utils/eventshandler.py]: The point is that it will matter for other implementations of `EventsHandler` so you need to be able to have an interface that is appropriately generic that it works for everything.

File: remotespark/utils/sparkevents.py (hunk 17 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: imports python class dynamically.

File: remotespark/utils/sparkevents.py (hunk 21 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: Have a loot at this: http://stackoverflow.com/questions/547829/how-to-dynamically-load-a-python-class

Comment (msftristew) [inline remotespark/utils/sparkevents.py]: Yes, but it's still an internal implementation detail that shouldn't be used by us:

> Import a module. Because this function is meant for use by the Python interpreter and not for general use it is better to use importlib.import_module() to programmatically import a module.

Comment (msftristew) [inline remotespark/utils/sparkevents.py]: `importlib` also has an easier API so you don't have to do the thing with the for loop and the `getattr`

Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: OK

File: remotespark/utils/eventshandler.py (hunk 15 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/eventshandler.py]: Do you mean "*args" or "**args"?

Comment (msftristew) [inline remotespark/utils/eventshandler.py]: Yes, `*args`, sorry.

File: remotespark/livyclientlib/livysession.py (hunk 14 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Yes, it should be a unique for every session created. It'll help us to identify all of the related events to specific session instead of using session Id which can be similar with other ones.

Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Actually, we can store the value without adding it to session state.

File: remotespark/utils/sparkevents.py [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+        Create an instance from the handler mentioned in the config file.
+        """
+        events_handler = importlib.import_module(conf.events_handler_class())
+        self.handler = events_handler()
+
+    def emit_start_session_event(self, time_stamp, event_name, session_guid, language):
+        """
+        Emitting Start Session Event
+        """
+        raise NotImplemented
```
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: This isn't valid Python code.

File: remotespark/utils/sparkevents.py [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+
+    def emit_end_session_event(self,time_stamp, event_name, session_guid, language, session_id):
```
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: Please watch your spaces in-between parameters

File: remotespark/livyclientlib/livysession.py [hunk 24]
```diff
@@ -50,9 +51,9 @@ def __init__(self, http_client, properties, ipython_display,
         self._status_sleep_seconds = status_sleep_seconds
         self._statement_sleep_seconds = statement_sleep_seconds
         self._create_sql_context_timeout_seconds = create_sql_context_timeout_seconds
+        self.session_guid = uuid.uuid4()
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: We should talk about GUIDs for events. There's a bunch of operations that will need GUIDs. We shouldn't pollute all of our methods...

I believe the GUID should be handled by `SparkEvents` behind the covers. Let's talk about it tomorrow.

File: remotespark/utils/sparkevents.py [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+
+class SparkEvents:
+    def __init__(self):
+        """
+        Create an instance from the handler mentioned in the config file.
+        """
+        events_handler = importlib.import_module(conf.events_handler_class())
+        self.handler = events_handler()
+
+    def emit_start_session_event(self, time_stamp, event_name, session_guid, language):
```
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: What's the dimensions for each of these events?

I believe you don't need to pass in the `event_name` or `time_stamp`. Both of these fields can be handled by this class.

File: remotespark/utils/eventshandler.py [hunk 26]
```diff
@@ -0,0 +1,13 @@
...
+class EventsHandler:
+    def __init__(self):
+        self.logger = Log("EventsHandler")
+
+    def store_event(self, *args):
+        """
+        Storing the Event details using the logger.
+        """
+        event_line = ",".join("{}".format(arg) for arg in args)
+        self.logger.debug(event_line)
```
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: Call `info` method instead.

File: remotespark/utils/eventshandler.py [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+
+class EventsHandler:
+    def __init__(self):
+        self.logger = Log("EventsHandler")
+
+    def store_event(self, *args):
+        """
+        Storing the Event details using the logger.
+        """
+        event_line = ",".join("{}".format(arg) for arg in args)
```
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: Should this be a key-value string?

File: remotespark/utils/eventshandler.py [hunk 28]
```diff
@@ -0,0 +1,13 @@
+from remotespark.utils.log import Log
+
+
+class EventsHandler:
+    def __init__(self):
+        self.logger = Log("EventsHandler")
+
+    def store_event(self, *args):
```
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: If this is a `handler`, should we name the method `handle_event`?

File: remotespark/utils/sparkevents.py (hunk 25 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: I was thinking about the same thing, this make more sense.

File: remotespark/utils/sparkevents.py (hunk 22 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: What do you mean?

File: remotespark/utils/eventshandler.py (hunk 28 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/eventshandler.py]: More meaningful. will change it.

Comment (MohamedElKamhawy) [inline remotespark/utils/eventshandler.py]: I see your point but the main job for this method is just storing the event info, what do you think?

File: remotespark/utils/eventshandler.py (hunk 27 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/eventshandler.py]: If we want to unify the API across the different handlers the list will be more fit. on the other side, key-value will give more sense to the logs information.  So, I go for the current state, what do you think?

File: remotespark/utils/eventshandler.py (hunk 28 above)
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: I prefer handle. @msftristew ?

File: remotespark/utils/eventshandler.py (hunk 27 above)
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: What does 
>unify the API across the different handlers

mean?

File: remotespark/utils/sparkevents.py (hunk 22 above)
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: Run it on a python terminal. That class doesn't exist and you are not instantiating an object.

File: remotespark/utils/eventshandler.py (hunk 27 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/eventshandler.py]: When I add the new handler "hdinsighteventshandler" the method there will be the same and it will help to insert them in the MDS table with the same arrangement. 

Let's say we have dim1, dim2..etc. arranged in this way then, we put them in a key-value, we will need to iterate over them to insert them in the right way. 

So, if they are already sorted in a list? so we will insert them easily without need to iterate over the dictionary to put the value for each attribute in the right place.

File: remotespark/utils/eventshandler.py (hunk 28 above)
Comment (msftristew) [inline remotespark/utils/eventshandler.py]: Agreed with @aggFTW 

If the name is `EventsHandler`, it should handle events.  It should store events if its name is `EventStorage`

File: remotespark/utils/eventshandler.py (hunk 26 above)
Comment (msftristew) [inline remotespark/utils/eventshandler.py]: Log doesn't have an `info` method -- remember when you wrote that class @aggFTW ? :)

File: remotespark/utils/eventshandler.py (hunk 28 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/eventshandler.py]: Okay.

File: remotespark/utils/sparkevents.py (hunk 22 above)
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: Please try to test code before pushing it

File: remotespark/utils/eventshandler.py (hunk 26 above)
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: :(

File: remotespark/livyclientlib/livysession.py (hunk 24 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: We agreed to make this a decorator.

File: remotespark/utils/eventshandler.py (hunk 27 above)
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: You could have sorted key,value pairs. Just don't use a dictionary. Or don't sort them and iterate over them; what's wrong with that?

I'm not focusing on implementation, I'm focusing on building a robust API that any system could use.

File: remotespark/utils/sparkevents.py (hunk 22 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: I think this is the reason for the error you are getting: http://stackoverflow.com/questions/15514593/importerror-no-module-named-when-trying-to-run-python-script

Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: ?

File: remotespark/utils/eventshandler.py (hunk 27 above)
Comment (msftristew) [inline remotespark/utils/eventshandler.py]: A list of key-value pairs would satisfy Alejandro's request, @MohamedElKamhawy

File: remotespark/utils/sparkevents.py (hunk 22 above)
Comment (msftristew) [inline remotespark/utils/sparkevents.py]: Mohamed, Alejandro is talking about how this should be

    raise NotImplementedError()

File: remotespark/utils/decorators.py [hunk 29]
```diff
@@ -0,0 +1,17 @@
...
+    def class_decorator(cls):
+        def get_attr(self, attr=attr_name):
+            return getattr(self, "_" + attr)
+
+        def set_attr(self, value, attr=attr_name):
+            setattr(self, "_" + attr, value)
+
+        prop = property(get_attr, set_attr)
+        setattr(cls, attr_name, prop)
+        setattr(cls, "_" + attr_name, utils.generate_uuid())  # Default value for that attribute
```
Comment (msftristew) [inline remotespark/utils/decorators.py]: This adds the same GUID to all sessions (as a class variable).  Two alternatives:

* Edit the class's `__init__` method directly in this decorator

* Eschew the decorator and just use a base class for all things with GUID's in them (I prefer this)

**Comment by MohamedElKamhawy** [comment]
@msftristew , @aggFTW  I have added a new PR, hopefully it's the last before adding the unit tests.

**Comment by MohamedElKamhawy** [comment]
If there is no more comments, I fixed the last comments and I'll push per merging.

File: remotespark/livyclientlib/sparkcontroller.py [hunk 30]
```diff
@@ -73,12 +74,17 @@ def add_session(self, name, connection_string, skip_if_exists, properties):
             return
 
         session = self._create_livy_session(connection_string, properties, self.ipython_display)
+
+        spark_events = SparkEvents()
```
Comment (msftristew) [inline remotespark/livyclientlib/sparkcontroller.py]: Why not create one `SparkEvents` object in the constructor and use it for all events on the `SparkController`?

**Comment by msftristew** [comment]
Mohamed, please do not merge anything until you have it thoroughly tested.

**Comment by MohamedElKamhawy** [comment]
@msftristew  Sure.

File: remotespark/livyclientlib/livysession.py [hunk 31]
```diff
@@ -12,7 +13,8 @@
 from .livyreliablehttpclient import LivyReliableHttpClient
 
 
-class LivySession(object):
+
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: Please revert all whitespace changes on this file. It's no longer PEP compliant.

File: remotespark/livyclientlib/sparkcontroller.py (hunk 30 above)
Comment (aggFTW) [inline remotespark/livyclientlib/sparkcontroller.py]: What about emitting the events right before we make the call to Livy and right when we get the call that says it's in running state? This includes hiveContext creation, which should be a different operation.

File: remotespark/utils/constants.py [hunk 32]
```diff
@@ -9,6 +9,10 @@
 SESSION_KIND_SPARKR = "sparkr"
 SESSION_KINDS_SUPPORTED = [SESSION_KIND_SPARK, SESSION_KIND_PYSPARK, SESSION_KIND_SPARKR]
 
+SESSION_CREATION_START_EVENT = "notebookSessionCreationStart"
+SESSION_CREATION_END_EVENT = "notebookSessionCreationEnd"
+EVENTS = {SESSION_CREATION_START_EVENT, SESSION_CREATION_END_EVENT}
```
Comment (aggFTW) [inline remotespark/utils/constants.py]: What is this `EVENTS` for?

File: remotespark/utils/eventshandler.py [hunk 33]
```diff
@@ -0,0 +1,13 @@
+from remotespark.utils.log import Log
+
+
+class EventsHandler:
+    def __init__(self):
+        self.logger = Log("EventsHandler")
+
+    def handle_event(self, args):
```
Comment (aggFTW) [inline remotespark/utils/eventshandler.py]: I'd rename `args` to reflect it's a key, value collection

File: remotespark/livyclientlib/livysession.py [hunk 34]
```diff
@@ -12,7 +13,8 @@
...
+class LivySession(GuidMixin, object):
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: Don't you have to call the parent's constructor `__init__` method for the mixin to work?

File: remotespark/utils/sparkevents.py [hunk 35]
```diff
@@ -0,0 +1,48 @@
...
+import remotespark.utils.configuration as conf
+import remotespark.utils.constants as constants
+
+
+class SparkEvents:
+    def __init__(self):
+        """
+        Create an instance from the handler mentioned in the config file.
+        """
+        module, class_name = conf.events_handler_class().rsplit('.', 1)
```
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: What if there's no `.` in the config?

File: remotespark/utils/sparkevents.py [hunk 36]
```diff
@@ -0,0 +1,48 @@
...
+        events_handler_module = importlib.import_module(module)
+        events_handler = getattr(events_handler_module, class_name)
+
+        self.handler = events_handler()
+
+
```
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: Make whitespace in the file PEP-compliant.

File: remotespark/utils/sparkevents.py [hunk 37]
```diff
@@ -0,0 +1,48 @@
...
+    def emit_session_creation_start_event(self, session_guid, language):
+        """
+        Emitting Start Session Event
+        """
+        assert language in constants.SESSION_KINDS_SUPPORTED
+
+        event_name = constants.SESSION_CREATION_START_EVENT
+        time_stamp = datetime.now()
```
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: Getting the current time should be in a method by itself. What is the timezone used by `datetime.now()`?

File: remotespark/utils/sparkevents.py [hunk 38]
```diff
@@ -0,0 +1,48 @@
...
+        """
+        assert language in constants.SESSION_KINDS_SUPPORTED
+        assert session_id >= 0
+
+        args = []
+
+        event_name = constants.SESSION_CREATION_END_EVENT
+        time_stamp = datetime.now()
+
+        args = [("TimeStamp", time_stamp), ("EventName", event_name), ("SessionGuid", session_guid),
```
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: What about the final state?

What metrics will we be able to collect from these events?

File: tests/test_sparkcontroller.py [hunk 39]
```diff
@@ -41,6 +42,8 @@ def test_add_session():
     connection_string = "url=http://location:port;username=name;password=word"
     client = MagicMock()
     session = MagicMock()
+    session.kind = SESSION_KIND_SPARK
```
Comment (aggFTW) [inline tests/test_sparkcontroller.py]: Why did you add this? Your unit tests should be emitting events to memory and should not be using the real implementation.

File: remotespark/utils/constants.py (hunk 32 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/constants.py]: For newly added events, this should be reflected here instead of adding it in the SparkEvents file.

File: remotespark/livyclientlib/livysession.py (hunk 34 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: No, no need for that. I tried it out myself and works fine.

File: remotespark/livyclientlib/sparkcontroller.py (hunk 30 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/sparkcontroller.py]: Yes, I agree with you @aggFTW  to be before the call to Livy but for session creation end I think this is the right place for it .

File: remotespark/utils/sparkevents.py (hunk 37 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: It returns Local date & time. I'll use the utc instead.

File: remotespark/utils/sparkevents.py (hunk 38 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: The intention for those two events is to get the Average Session Creation Time.

File: remotespark/utils/sparkevents.py (hunk 35 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: Anyone should use his own handler should add the whole path till the class used as the handler. Else, this will not work (Compilation Error) as we are using importlib_module method which expects module.

File: remotespark/utils/constants.py (hunk 32 above)
Comment (aggFTW) [inline remotespark/utils/constants.py]: Sorry, didn't get it. Who is going to use this variable?

File: remotespark/livyclientlib/sparkcontroller.py (hunk 30 above)
Comment (aggFTW) [inline remotespark/livyclientlib/sparkcontroller.py]: That's not true. We can get both before the call to Livy.

File: remotespark/utils/constants.py (hunk 32 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/constants.py]: This will contain all of our added events to save them in one place.

Comment (msftristew) [inline remotespark/utils/constants.py]: Why?

File: remotespark/livyclientlib/livysession.py (hunk 34 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: From what I read, you DO need to call it, no?
http://stackoverflow.com/a/1385836

File: remotespark/utils/constants.py (hunk 32 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/constants.py]: You suggested that Ricky, remember?

File: remotespark/utils/sparkevents.py (hunk 38 above)
Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: And what else? That's one of N metrics we want for session creation. Will the schema support all the metrics? For a list of metrics, refer to the email I shared with a list of metrics from Asad.

File: tests/test_sparkevents.py [hunk 40]
```diff
@@ -0,0 +1,43 @@
...
+def test_emit_session_creation_start_event():
+
+    language = constants.SESSION_KIND_SPARK
+    event_name = constants.SESSION_CREATION_START_EVENT
+
+    kwargs_list = [("TimeStamp", spark_events.get_utc_date_time()), ("EventName", event_name), ("SessionGuid", guid),
+                ("SparkLanguage", language)]
+
+    spark_events.emit_session_creation_start_event(guid, language)
+    spark_events.get_utc_date_time.assrt_called_once_with()
```
Comment (aggFTW) [inline tests/test_sparkevents.py]: Typo in assert.

File: remotespark/livyclientlib/livysession.py (hunk 34 above)
Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: Mohamed, the `__init__` method does not get called automatically.

hdiuser@hn0-jupyte:~$ python
Python 2.7.6 (default, Jun 22 2015, 17:58:13)
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> class A(object):
...     def __init__(self):
...             print 'yo'
...
>>> class B(object):
...     def __init__(self):
...             print 'wat'
...
>>> b = B()
wat
>>>

File: remotespark/utils/constants.py (hunk 32 above)
Comment (msftristew) [inline remotespark/utils/constants.py]: No.

File: remotespark/livyclientlib/livysession.py (hunk 34 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: How did you try it and had it work Mohamed? I'd be very interested in seeing how you got it to work without calling it explicitly.

File: remotespark/utils/constants.py (hunk 32 above)
Comment (msftristew) [inline remotespark/utils/constants.py]: The question is, why are you creating a set with all the events in it?

File: remotespark/utils/sparkevents.py (hunk 38 above)
Comment (MohamedElKamhawy) [inline remotespark/utils/sparkevents.py]: Yes, I believe that. We have discussed that when I sent the design document. We have 11 dimensions, that can be used to satisfy the different metrics mentioned in the email you have sent. For now, let's start with Average Session creation Time as Sangeetha wanted then will add the new metrics gradually as needed.

Comment (aggFTW) [inline remotespark/utils/sparkevents.py]: Talked f2f. You'll go over the document and let us know if we need to log any more information with these events for other metrics so that we don't have breaking changes in the information we are logging for each event.

Please let us know the outcome of your investigation before you merge.

**Comment by MohamedElKamhawy** [comment]
@msftristew , @aggFTW  Can you have a look? It seems that it didn't notify you yesterday with the new commit.

**Comment by msftristew** [comment]
GitHub never notifies us about new commits, please notify us if you have any updates we need to be aware of

**Comment by MohamedElKamhawy** [comment]
@msftristew  Yes, please check the new commit from yesterday.

**Comment by MohamedElKamhawy** [comment]
I have added a new commit, it'll be great if you can give me the whole comments you see to speed the work at my side.

**Comment by aggFTW** [comment]
I cannot give comments on new changes introduced by commits before the commit is made :)

File: remotespark/livyclientlib/livysession.py (hunk 34 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Here is the code, I tried:

class GuidMixin(object):
    def __init__(self):
        self.guid = generate_uuid()

class Subclass(GuidMixin):

    def printx(self):
        print self.guid

if __name__ == "__main__":
    a = Subclass()
    b = Subclass()

    print a.guid
    a.printx()

    print b.guid
    b.printx()

**Comment by aggFTW** [comment]
LGTM!

**Comment by msftristew** [comment]
Just rename `GuidMixin` to something else (it's not really a mixin) and then LGTM

**Comment by msftristew** [comment]
Is this tested E2E? You've run Jupyter and verified the events are properly logged?
